rpc: Round verificationprogress to 1 for a recent tip

maflcko opened the pull request:
Some users really seem to care about this. While it shouldn't matter much, the diff is so trivial that it is probably worth doing.

Fixes #31127

One could also consider to split the field into two dedicated ones (https://github.com/bitcoin/bitcoin/issues/28847#issuecomment-1807115357), but this is left for a more involved follow-up and may also be controversial.

Commit:
test: Use mockable time in GuessVerificationProgress

Also add a test.

Commit:
rpc: Round verificationprogress to exactly 1 for a recent tip

This requires a new lock annotation, but all relevant callers already
held the lock.

pinheadmz reviewed (approved):
ACK 11116aa19f5173792f757276c8a58b279f18a199

Built and tested on macos/arm64. Reviewed the code. Grudgingly I accept that this is a user complaint that will never go away and since the PR only updates the UI I'm ok with it after all.

`guessVerificationProgress()` will now return `1` if the timestamp in the chain tip is within the last two hours. (Probably should have a release note?)

That gives the following result in the log when I finished syncing my mainnet full node (at about 13:40:00 UTC):

```
height=896951 date='2025-05-16T09:17:23Z' progress=0.999967
height=896952 date='2025-05-16T09:27:05Z' progress=0.999969
height=896953 date='2025-05-16T09:57:42Z' progress=0.999976
height=896954 date='2025-05-16T10:02:10Z' progress=0.999977
height=896955 date='2025-05-16T10:13:23Z' progress=0.999980
height=896956 date='2025-05-16T10:14:26Z' progress=0.999980
height=896957 date='2025-05-16T10:18:47Z' progress=0.999981
height=896958 date='2025-05-16T10:42:15Z' progress=0.999986
height=896959 date='2025-05-16T11:52:58Z' progress=1.000000
height=896960 date='2025-05-16T11:54:12Z' progress=1.000000
height=896961 date='2025-05-16T11:59:16Z' progress=1.000000
height=896962 date='2025-05-16T11:59:17Z' progress=1.000000
height=896963 date='2025-05-16T12:10:47Z' progress=1.000000
height=896964 date='2025-05-16T12:13:21Z' progress=1.000000
height=896965 date='2025-05-16T12:38:33Z' progress=1.000000
height=896966 date='2025-05-16T12:38:43Z' progress=1.000000
height=896967 date='2025-05-16T12:45:46Z' progress=1.000000
height=896968 date='2025-05-16T12:47:09Z' progress=1.000000
height=896969 date='2025-05-16T13:07:02Z' progress=1.000000
height=896970 date='2025-05-16T13:19:45Z' progress=1.000000

```

The only other place the "guess" is called is during wallet rescan. I tried to fill up a wallet with descriptors and start a rescan to try and catch a "progress: 100%" in the logs but couldn't nail it after a few attempts, which is a good thing. Those log messages look like this:

- `Still rescanning. At block 896884. Progress=0.999814`
- `Rescan interrupted by shutdown request at block 896735. Progress=0.999468`

I also tried a rescan in the GUI but its just a progress bar, no integer percentage and the last ~12 blocks are processed fast enough that I think even seeing a complete-looking status for a few extra seconds is worth that glowing smile on the user's face ;-)

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 11116aa19f5173792f757276c8a58b279f18a199
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgnRqEACgkQ5+KYS2KJ
yToGIQ//Ysxl/euuLbz/Wq/F2abHwOb/jOcKVHe8kZjtYokr2yCENPWA+RfqNtKP
KKJZbsilvUa8YAFl683opJj0m92M7G+dU3m2Opq/+VwuVlrfF8AHHxYgV2MK3E+U
csK4QjeHBrXsWsMKv31sEL7ogG/oQ67EMiuFPjTzQpNsnZDiRUPwRE4bSefmQ4Jw
eS2m/Jp3bVxCjCweFamsUVg/WtNeIDPcxa95cIeY9WarWOsFqmZR2wGsirICDVvj
K74AD6zq5YV7jMI8DVcYHGfd95/stydV+J9CA3XVeQQDHv1NtVcZQBGFpubT7k4O
KlUVUKmFF1TfExAwfJYgY98u6NoXQ811XvTQgZh9WC1OmxWucyUdpVWsZbMMdFz2
MhA46Z5lpgb23rPW57+U/MzczMFWKSS9jWCDF/9GA0PwZnox+73Wlrtk/A+Sgs1D
HSv+6OYRrrw19rbzflEz1KN+ON0jxBVrtW5kl0xNmZmrJO5ON9jJS0aojZTT3kDT
2ON34yQjS9ySFVsxPQdsQ+JuNd0mfK/DZz0v5VbirEp/gSv3xczHDkfxSQz9cldY
m01cq+jIaczpCKSvIQneO61f+Pfw4SixQm3TUv3NVSte4syAU811pyC74NLBxtDj
qwMimuHlydKd4D0jPbqO3/lj85b6nuytIIenResP5aZmiLuh9f8=
=Ub9B
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

l0rinc commented:
Will this show `progress=1.000000` before the last block as well?
I find that more confusing than `0.999999` for the last one.

l0rinc commented:
So what does the percentage mean if there's no fixed target value in the first place?

pinheadmz commented:
After this PR, it will mean your chain tip's timestamp is within 2 hours of now.

jonatack commented:
Concept ACK

polespinasa reviewed (commented):
cACK

jasonribble commented on src/validation.cpp:
```diff
@@ -5572,14 +5572,21 @@ double ChainstateManager::GuessVerificationProgress(const CBlockIndex* pindex) c
+    // verificationprogress of exactly "1.0" once they are close to the tip.
+    // The progress is clamped before returning to at most "1.0", even if the
+    // miner set a timestamp off from the real time.
+    const auto block_time{pindex->GetBlockTime() + Ticks<std::chrono::seconds>(2h)};
```
What's the benefit of doing it this way? /curi

polespinasa replied:
> What's the benefit of doing it this way? /curi

By checking if the last block verified is the one matching the last best header known, we avoid previous blocks inside the 2h offset return `verificationprogress=1`.

jasonribble replied:
Good moves. I like ternary.

jasonribble replied:
But doesn't the current code already solve the problem?

polespinasa replied:
> But doesn't the current code already solve the problem?

No, maflcko proposal returns `verificationprogress=1` when the last verified block is within a 2h offset of our local time.

jasonribble commented:
cACK

Commit:
refactor: Pass verification_progress into block tip notifications

It is cheap to calculate and the caller does not have to take a lock to
calculate it.

Also turn pointers that can never be null into references.

polespinasa commented on src/validation.cpp:
```diff
@@ -5572,14 +5575,25 @@ double ChainstateManager::GuessVerificationProgress(const CBlockIndex* pindex) c
+            // close to "1.0", because some users expect it to be. This also
+            // avoids relying too much on the exact miner-set timestamp, which
+            // may be off.
+            nNow - (m_best_header->nHeight - pindex->nHeight) * GetConsensus().nPowTargetSpacing :
```
I guess `GetConsensus().nPowTargetSpacing` returns 10min?

If so, could we use this even if the block being validated is not within the 2h range? I think we could just never rely on miner-set timestamps.

polespinasa reviewed (commented):
tACK fa53098472521de9d5b3cc42983043c240b7c321

Note: The failing test `rpc_signer.py ` is working for me locally

Left a small comment

pinheadmz commented on src/validation.cpp:
```diff
@@ -3766,6 +3767,9 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
         }
 
         InvalidChainFound(to_mark_failed);
+        if (pindex_was_in_chain) {
+            failed_prev_verification_progress = m_chainman.GuessVerificationProgress(to_mark_failed->pprev);
+        }
```
fa53098472521de9d5b3cc42983043c240b7c321

Why is this here and not in the next block that is already guarded by `pindex_was_in_chain`? `to_mark_failed` and `m_chainman` are still in scope.

pinheadmz commented:
One thing that's a little weird about this is that when I open regtest after a few days...

```
--> bccli -regtest -getinfo
Chain: regtest
Blocks: 202
Headers: 202
Verification progress: ▒▒▒▒▒▒░░░░░░░░░░░░░░░ 27.7195%
```

Of course after generating 1 block at time=now, it snaps to 100%

davidgumberg commented on src/node/interfaces.cpp:
```diff
@@ -409,6 +410,7 @@ class NodeImpl : public Node
     std::unique_ptr<Handler> handleNotifyBlockTip(NotifyBlockTipFn fn) override
     {
         return MakeSignalHandler(::uiInterface.NotifyBlockTip_connect([fn, this](SynchronizationState sync_state, const CBlockIndex* block) {
+            LOCK(chainman().GetMutex());
```
Am I correct in understanding that `handleNotifyBlockTip` takes a lock now where it didn't before?

pinheadmz commented:
Also this is somewhat pathalogical but...

```

--> bccli -regtest setmocktime 1719861124 # July 1, 2024
--> bccli -regtest -getinfo
Chain: regtest
Blocks: 203
Headers: 203
Verification progress:  -0.7451%

```

polespinasa commented:
> when I open regtest after a few days...

Maybe we should force this code to work only in mainnet? It doesn't make sense the time considerations for regtest. Maybe testnet and signet could still use it.

pinheadmz commented:
Approach ACK fa53098472 (still reviewing code changes around blocktip notifications)

With fresh brain this morning I think this is the best possible approach if we make any change at all. The regtest stuff I observed is expected behavior - if there's no block for 2 days then yes you are 288 blocks behind and progress should estimate that. The new chainparams define regtest as having an expected 0.001 tx/s which is approximately equal to one (coinbase) tx every ten minutes.


Example catching up:

```

date='2025-05-22T12:45:54Z' progress=0.999979
date='2025-05-22T12:47:48Z' progress=0.999981
date='2025-05-22T12:51:57Z' progress=0.999984
date='2025-05-22T13:01:53Z' progress=0.999986
date='2025-05-22T13:17:07Z' progress=0.999988
date='2025-05-22T13:31:46Z' progress=0.999991
date='2025-05-22T13:39:06Z' progress=0.999993
date='2025-05-22T13:46:58Z' progress=0.999995
date='2025-05-22T13:56:12Z' progress=0.999998
date='2025-05-22T14:01:13Z' progress=1.000000
```

pinheadmz reviewed (approved):
ACK fa53098472521de9d5b3cc42983043c240b7c321

Tested and reviewed code. Might warrant a release note: "Verification progress will be 100% if all known blocks are verified and timestamped within the last two hours when previously a stale tip might result in progress < 100%".

In addition to this UI change, block tip signal callers now compute the estimated progress as opposed to the signal handlers, and `CBlockIndex` are being passed by reference instead of raw pointer through the signal-handler chain.


<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK fa53098472521de9d5b3cc42983043c240b7c321
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmgvclwACgkQ5+KYS2KJ
yTrUsA//QL0jGfEKqsy8PhRtfS2CqyFQMS+VTDeSeBofJHv0El1Tuwq5fMiLNG5U
O3twA1SIvCSGJOjFlsDxvgb4PnxqnR5ioHH31aN1uZK1+wN/GW5k8mDYg5ul2kWP
nfZ031w/z93yftyuB+APqqKaZsdVzhVKzw97yscvaPCIOO8kzUN1DoQySfVX6L/V
WVHGaAbXk8eE1qWReWeux/AWy8PwDZMjCfNZkWggpexFyRPVNn8JFPGvG7xrUjWe
4qtsrK3juuw9UljForsm2ibGe7QtXw7jTPh/OTCyYQvJoNDgN/2/Z9I00TI8l2bK
TX9DpXUXyxDglEDRvDXybHmmPJrsJquDvqQPIr6zv//PsSM/WKs6jwX5p0J8DB5b
v+cWGstX4rlvlEWHx5djk8+NKL3Eb377kq/Rn2EpNXpvtsLvlTIG4OS3TkmkTReX
WGvFfbGabk+CxeD0Sntg5pmmDMdFem2YRIekasmjB7M5AJQrn51YL8PyeMaseQHG
EBToUfNzXR4o3pJtorso3cIG5QTAeL9yvvQC7+VfmFyxYLv9TOThMBROHvabjotr
a4H9GPITfJwBluKfA6qqgFg2j1g2bOLB6gpkvzwGFTQI83O+myQxZexmvUzD419b
C6uMVolX85w26iTY+YuUB6cCXzkl4/O21vj6Oe1Wqgy0N7QDVOo=
=qaQr
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

polespinasa commented:
> One thing that's a little weird about this is that when I open regtest after a few days...
> 
> ```
> --> bccli -regtest -getinfo
> Chain: regtest
> Blocks: 202
> Headers: 202
> Verification progress: ▒▒▒▒▒▒░░░░░░░░░░░░░░░ 27.7195%
> ```
> 
> Of course after generating 1 block at time=now, it snaps to 100%

Maybe instead of not using this new approach on regtest (as I suggested here https://github.com/bitcoin/bitcoin/pull/32528#issuecomment-2899113774) we could just add a message so users can understand why this happens? 


The actual result can be really confusing if you don't know how the % is calculated.

polespinasa commented:
> I fail to see how showing a value lower than 1.0 on regtest if your tip is stale is "really confusing".

Because normally you don't have a regtest environment running 24/7 mining blocks. You turn on and off the environment for testing. The use of timestamps doesn't make sense, there is no "possible not known" blocks to take into account.

polespinasa reviewed (commented):
tACK fab1e02086ceebd7d96417b7a386fe61158bfda9

The test added faf6304bdfdf228354b4072b72f4c0ef90fdaade lgtm and makes much sense.

(non-blocking) My only point is that a warning message at least would be good for regtest as being at the tip and having the verification progress < 1 when no new blocks are getting mined (in regtest) may confuse.

> Regtest is only used for testing and heavily mockable, so one should be cautious to put too much meaning into a value that even on mainnet is informational-only at best.

I agree but still informing the user is not a bad idea.

<details>
<summary>see proposal
</summary>

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 880e35ee7f..b694171583 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -1379,7 +1379,17 @@ RPCHelpMan getblockchaininfo()
     }
 
     NodeContext& node = EnsureAnyNodeContext(request.context);
-    obj.pushKV("warnings", node::GetWarningsForRpc(*CHECK_NONFATAL(node.warnings), IsDeprecatedRPCEnabled("warnings")));
+    UniValue warnings = node::GetWarningsForRpc(*CHECK_NONFATAL(node.warnings), IsDeprecatedRPCEnabled("warnings"));
+    if (chainman.GetParams().GetChainType() == ChainType::REGTEST) {
+        const std::string regtest_warning = "Verification progress values may not reach 1 until a new block is mined";
+        if (IsDeprecatedRPCEnabled("warnings")) {
+            if (warnings.isStr()) warnings = warnings.get_str() + "; " + regtest_warning;
+        } else {
+            warnings.push_back(regtest_warning);
+        }
+    }
+
+    obj.pushKV("warnings", warnings);
     return obj;
 },
     };
```
</details>

pinheadmz reviewed (approved):
ACK fab1e02086ceebd7d96417b7a386fe61158bfda9

Trivial code cleanup changes since last ack. Also re-built and tested on macos/arm64

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK fab1e02086ceebd7d96417b7a386fe61158bfda9
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmg1wvoACgkQ5+KYS2KJ
yTrrtBAAhzspf6yfD6do0zH7Cu0I4OMA1C/KlabXxcyA9O4Foa/sNjx//iXEKFIX
EO0fEfcGZ+cWwlpMGG5TD+NfPl6jKfVb376xxESSUT4HvUIOB//U8nDi3cy4LkBH
iYz7qpieDCuOi+3sLZeAT7aFRlnJSr8p8IzzmkAIAOhkZbKNtrhAa6Qzx3/ax6SM
C0F7sxQdpCeeKUaqmqWL+JJRztSKK/ohwKsvQLZ+uUQAiG+5AV70ycFVNueFqS12
0Ysz0jaKMQaSuL4lXe36699tIEb+Ep8o4/qAF5h4itRZa7zy2/aFC7gSd3/rFCLD
ErCLqjjkhPYdNJHZ+Gc/UloKAUPHS7BC9XKpsmM9gOBTtH17jKycL3zT+CazSdNx
INh6rbaqelp+UXDqNhYdxyEqYpW8siVrgT62a2CU65WpV/1J1/OyJB4jgOch507Q
2QW19jkrKURISbqtZTX8locKPW4eAhEahxAMSJC5fHZa2seRfnEVpUN0T90TaGx3
ol6XygxQJiFcnu6HdMzJ3fmdT6vStYxxpvyoJOD4e3+JT6x2GtHhaL7DiY+Ax3FM
6pHCyMJehC67w9LnKXTQ11OoGij2ajQ5J34xARCcgl3XSQdkCZOcFzjw3g3Cdfiw
N/S8vFdzBJEGhysnIAVrtEkzdUjyXlixs4+BApT7wx9N58hrD5s=
=letv
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on openpgp.org](https://keys.openpgp.org/vks/v1/by-fingerprint/E61773CD6E01040E2F1BD78CE7E2984B6289C93A)
</details>

sipa reviewed (commented):
utACK fab1e02086ceebd7d96417b7a386fe61158bfda9

achow101 commented:
ACK fab1e02086ceebd7d96417b7a386fe61158bfda9
